Kernel/Thread.cpp

```diff
@@ -710,56 +708,107 @@ DispatchSignalResult Thread::dispatch_signal(u8 signal)
-        u32 ret_eip = state.eip;
-        u32 ret_eflags = state.eflags;
+    // We now place the thread state on the userspace stack.
+    // Note that we use a RegisterState.
+    // Conversely, when the thread isn't blocking the RegisterState may not be
+    // valid (fork, exec etc) but the tss will, so we use that instead.
+    auto& regs = get_register_dump_from_stack();
+
+    setup_signal_stack(regs, signal, old_signal_mask, handler_vaddr);
+    regs.eip = g_return_to_ring3_from_signal_trampoline.get();
 
 #ifdef SIGNAL_DEBUG
-        klog() << "signal: setting up user stack to return to eip: " << String::format("%p", ret_eip) << " esp: " << String::format("%p", old_esp);
+    klog() << "signal: Okay, " << *this << " {" << state_string() << "} has been primed with signal handler " << String::format("%w", m_tss.cs) << ":" << String::format("%x", m_tss.eip) << " to deliver " << signal;
 #endif
+    return DispatchSignalResult::Continue;
+}
 
-        // Align the stack to 16 bytes.
-        // Note that we push 56 bytes (4 * 14) on to the stack,
-        // so we need to account for this here.
-        u32 stack_alignment = (*stack - 56) % 16;
-        *stack -= stack_alignment;
+void Thread::setup_signal_stack(RegisterState& state, u8 signal, u32 old_signal_mask, VirtualAddress handler_vaddr)
+{
+    u32* stack = &state.userspace_esp;
+    u32 old_esp = *stack;
+    u32 ret_eip = state.eip;
+    u32 ret_eflags = state.eflags;
 
-        push_value_on_user_stack(stack, ret_eflags);
+#ifdef SIGNAL_DEBUG
+    klog() << "signal: setting up user stack to return to eip: " << String::format("%p", ret_eip) << " esp: " << String::format("%p", old_esp);
+#endif
 
-        push_value_on_user_stack(stack, ret_eip);
-        push_value_on_user_stack(stack, state.eax);
-        push_value_on_user_stack(stack, state.ecx);
-        push_value_on_user_stack(stack, state.edx);
-        push_value_on_user_stack(stack, state.ebx);
-        push_value_on_user_stack(stack, old_esp);
-        push_value_on_user_stack(stack, state.ebp);
-        push_value_on_user_stack(stack, state.esi);
-        push_value_on_user_stack(stack, state.edi);
+    // FIXME: Is there a way to avoid putting the extra 60 bytes for SIGINFO
+    //        onto every single signal handler stack, even when siginfo was not specified?
 
-        // PUSH old_signal_mask
-        push_value_on_user_stack(stack, old_signal_mask);
+    // FIXME: Plumb sender information into this part. We need to know more about the signal.
+    //        Maybe need a struct SignalReason?
+    auto signal_info = siginfo_for_signal(signal, nullptr);
 
-        push_value_on_user_stack(stack, signal);
-        push_value_on_user_stack(stack, handler_vaddr.get());
-        push_value_on_user_stack(stack, 0); //push fake return address
+    ucontext_t user_context {};
+    user_context.uc_link = nullptr; // TODO: Are we expected to maintain a linked list of these things in the kernel? in userspace?
```

Comment:
This guy is for `swapcontext()`, it's set up by the user after `makecontext()` to be the context to return to when the swapped context returns.

Reply:
see: https://pubs.opengroup.org/onlinepubs/7908799/xsh/getcontext.html and https://pubs.opengroup.org/onlinepubs/7908799/xsh/makecontext.html

Reply:
Weird. So it's like old-style coroutines? That have to go through the kernel? And nobody in their right mind would use anymore? Or would the user context in a signal handler not include any kind of context data, and the save/return stuff would all be handled by the C library?

Reply:
That, I don't know for sure - no one (right mind or not) uses that stuff.
re in the signal handler, I think you just get the context as `getcontext()` would give you, of wherever the thread was. not sure if this is used in the resumption though
I think this whole `swapcontext()` is just a libc thing, the kernel isn't involved here (again, I think), afaik it just sets up the registers, pushes an address and then `ret`'s to it. not entirely sure though?

Reply:
> SUSv2, POSIX.1-2001.  POSIX.1-2008 removes the specifications of
> makecontext() and swapcontext(), citing portability issues, and
> recommending that applications be rewritten to use POSIX threads
> instead.

Ah yes. pre-pthread "threads". Interesting that they were actually removed. Probably a better use of time to implement C++20 coroutine support than those crusty APIs.

Reply:
I'm all for implementing C++20 coros over `swapcontext()` :P